colexec: check that Init is always called before Next and DrainMeta

yuzefovich opened the pull request:
**colexec: initialize operators in an edge case**

Previously, it was possible that `Init` is not called on some operators
but `DrainMeta` is. In particular, it was the case if we planned
`fixedNumTuplesNoInputOp` operator which replaced the whole subtree, and
that subtree contained some metadata sources. The assumption that all
operators in use are `Init`ialized was broken, so this commit updates
that utility operator to also initialize the subtree even though it
doesn't consider it as its input.

I noticed this when I was looking into the parallel unordered
synchronizer, and there is an open question whether the synchronizer
should perform its "full" initialization (i.e. spawn up the reader
goroutines) or whether the synchronizer should transition into "done"
state if `DrainMeta` is called before `Next`. The follow-up commit will
implement the latter.

Release justification: low-risk fix to existing functionality.

Release note: None

**colexec: check that Init is always called before Next and DrainMeta**

This commit adds another invariant to be checked - that `Init` is called
before `Next` and `DrainMeta`. It also adds the planning of the
invariants checker on top of the input operators (like synchronizers) and
updates the decision to plan the invariants checker to be made based on
`crdb_test` build tag (thus, extending its usage). The invariants checker
itself is moved into `colexec` package.

Release justification: non-production code change.

Release note: None

yuzefovich committed:
colexec: initialize operators in an edge case

Previously, it was possible that `Init` is not called on some operators
but `DrainMeta` is. In particular, it was the case if we planned
`fixedNumTuplesNoInputOp` operator which replaced the whole subtree, and
that subtree contained some metadata sources. The assumption that all
operators in use are `Init`ialized was broken, so this commit updates
that utility operator to also initialize the subtree even though it
doesn't consider it as its input.

I noticed this when I was looking into the parallel unordered
synchronizer, and there is an open question whether the synchronizer
should perform its "full" initialization (i.e. spawn up the reader
goroutines) or whether the synchronizer should transition into "done"
state if `DrainMeta` is called before `Next`. The follow-up commit will
implement the latter.

Release justification: low-risk fix to existing functionality.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61317)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: check that Init is always called before Next and DrainMeta

This commit adds another invariant to be checked - that `Init` is called
before `Next` and `DrainMeta`. It also adds the planning of the
invariants checker on top of the input operators (like synchronizers) and
updates the decision to plan the invariants checker to be made based on
`crdb_test` build tag (thus, extending its usage). The invariants checker
itself is moved into `colexec` package.

Release justification: non-production code change.

Release note: None

yuzefovich commented:
I think we should merge the first commit as is and backport it to 21.1 (possibly 20.2 as well), and then address #61679 more holistically (which will remove the special casing of the columnarizer removed from the flow).

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61317#-:-MVM9XNv6RM__19J2QT7:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/invariants_checker.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61317#-MVM9AWi73isMdxOBsiI:-MVM9AWi73isMdxOBsiJ:bhtd45d) ([raw file](https://github.com/cockroachdb/cockroach/blob/275b78968697a3f1f263ac3b2553c0256e6770bb/pkg/sql/colexec/invariants_checker.go#L56)):*
> ```Go
> 
> // Next implements the colexecop.Operator interface.
> func (i *InvariantsChecker) Next(ctx context.Context) coldata.Batch {
> ```

Might as well put a check in here as well.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61317#-:-MVNUyp88dhmzPRKOR-5:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/invariants_checker.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61317#-MVM9AWi73isMdxOBsiI:-MVNUxBC7T02iDB9mGMN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/275b78968697a3f1f263ac3b2553c0256e6770bb/pkg/sql/colexec/invariants_checker.go#L56)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Might as well put a check in here as well.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
